src/request/swap-iframe/SwapIFrameApi.js

```diff
@@ -573,6 +688,46 @@ class SwapIFrameApi extends BitcoinRequestParserMixin(RequestParser) { // eslint
             };
         }
 
+        if (parsedRequest.redeem.type === 'USDC' && storedRequest.redeem.type === 'USDC') {
+            const usdcHtlcContract = new ethers.Contract(
+                CONFIG.USDC_HTLC_CONTRACT_ADDRESS,
+                PolygonContractABIs.USDC_HTLC_CONTRACT_ABI,
+            );
+
+            // Place contract details into existing function call data
+            storedRequest.redeem.request.data = usdcHtlcContract.interface.encodeFunctionData(
+                storedRequest.redeem.description.name,
+                [
+                    /* bytes32 id */ parsedRequest.redeem.htlcId,
+                    /* address target */ storedRequest.redeem.description.args.target,
+                    ...(storedRequest.redeem.description.name === 'redeem' ? [
+                        /* uint256 approval */ storedRequest.redeem.description.args.secret,
+                    ] : []),
+                    /* uint256 fee */ storedRequest.redeem.description.args.fee,
+                ],
+            );
+
+            const typedData = new OpenGSN.TypedRequestData(CONFIG.POLYGON_CHAIN_ID, CONFIG.USDC_HTLC_CONTRACT_ADDRESS, {
+                request: storedRequest.redeem.request,
+                relayData: storedRequest.redeem.relayData,
+            });
+
+            const { EIP712Domain, ...cleanedTypes } = typedData.types;
+
+            const wallet = new ethers.Wallet(privateKeys.usdc);
+
+            const signature = await wallet._signTypedData(
+                typedData.domain,
+                /** @type {Record<string, ethers.ethers.TypedDataField[]>} */ (/** @type {unknown} */ (cleanedTypes)),
+                typedData.message,
+            );
```

Comment:
Everything is checked in the Hub's `SetupSwapSuccess`: the correct recipient address, the hash(Roots) and also the timeouts.

Reply:
I don't think that this solves the problem described above, and is also not entirely true.

The Hub receives the USDC HTLC info from Fastspot and forwards it to the Keyguard.
- Both, the Hub and Keyguard check whether the hash specified in that info matches the other swap contract.
- The Hub additionally checks whether the HTLC info has the expected redeem address or refund address. The Keyguard doesn't but should.
- The Keyguard to some degree checks the timeouts. The Hub doesn't but should.

But all that doesn't really matter because
- neither the Hub nor the Keyguard check the HTLC info against the actual HTLC contract data `htlcData`.
- the HTLC data is not contained in the data the Keyguard actually signs, as only the `htlcId` is contained in the signature.

I.e. the HTLC details, the HTLC contract data and the HTLC id can be completely unrelated and Fastspot could arbitrarily lie to us about what info we're signing and the details it claims we're signing, which neither the Hub nor the Keyguard check.

I don't know how the HTLC id is computed and whether we have all the data required to compute it. If we do, obviously the correct thing to do would be to recalculate it and check whether it matches. If we can't verify it, then the only thing we can do is checking whether the contract that Fastspot set-up on the chain actually contains what we expect (hash, recipient, timeout, etc.) during execution of the swap in the Wallet, before we send out our funding transaction, which comes back to the checks in libswap I already mentioned in the opening post.

It is actually a general problem that we do not or can not verify the HTLC id or address.
- For Nimiq, we can not verify the HTLC address, because the computation of the contract address includes the Fastspot sender address and validity start height which we don't now. However, libswap checks whether Fastspot's incoming funding transaction includes the correct htlc contract data, which was previously checked by the Keyguard. By this, the correctness of the HTLC address for the HTLC data is verified.
- For Bitcoin, the Keyguard only receives the UTXO of the HTLC funding transaction (transaction hash and output index), which the Hub fetched from the chain for the HTLC address, but which the Keyguard can not verify without fetching the chain itself. The HTLC address should be possible to be computed from the HTLC script to check that the HTLC address we got from Fastspot is correct, but the Hub currently doesn't do so. Neither does libswap, thus for Bitcoin, the correctness of the HTLC address for the HTLC data is _not_ verified.
- Fot USDC, as explained above, the correctness of the HTLC id is not verified in any way in the Keyguard, Hub or libswap. We would need to verify it in the Keyguard or Hub by recalculating it if possible, or otherwise in libswap by checking for Fastspot's incoming funding transaction to include the correct redeem address, hash and timeout.

Not verifying the correctness undermines the integrity of the atomic swap and would allow Fastspot to steal the user's funds by sending him the proper HTLC info the Hub and Keyguard would expect and verify, but an unrelated HTLC address or id for a HTLC the user can't claim. If the user then funds his part of the HTLC, Fastspot can claim his funds without the user being able to claim the redeem transaction. After the timeout, Fastspot would then be able to claim its own HTLC back, effectively stealing the user's funds.